File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        l1 = headA
        l2 = headB
        countA = 0
        countB = 0

        while l1:
            countA += 1
            l1 = l1.next
        while l2:
            countB += 1
            l2 = l2.next
        final = abs(countA - countB)
        
        if countA > countB:
            while final:
                headA = headA.next
                final -= 1
        if countB > countA:
            while final:
                headB = headB.next
                final -= 1
        while headA and headB:
            if headA == headB:
                return headA
            headA = headA.next
            headB = headB.next
        return None

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        HashSet<ListNode> hashset = new HashSet<>();
        
        while(headA != null) {
            hashset.add(headA);
            headA = headA.next;
        }
        while(headB != null) {
            if(hashset.contains(headB)) {
                return headB;
            }
            headB = headB.next;
        }
        
        return null;
    }
}